tools/oeedger8r/Emitter.ml

```diff
@@ -967,274 +472,861 @@ let warn_size_and_count_params (fd : func_decl) =
-          f.uf_fdecl.fname cconv_str ) ;
-      if f.uf_fattr.fa_dllimport then
-        failwithf "Function '%s': dllimport is not supported by oeedger8r."
-          f.uf_fdecl.fname ;
-      if f.uf_allow_list != [] then
-        printf
-          "Warning: Function '%s': Reentrant ocalls are not supported by Open \
-           Enclave. Allow list ignored.\n"
-          f.uf_fdecl.fname ;
-      if f.uf_is_switchless then
-        failwithf
-          "Function '%s': switchless ecalls and ocalls are not yet supported \
-           by Open Enclave SDK."
-          f.uf_fdecl.fname )
-    ec.ufunc_decls ;
-  (* Map warning functions over trusted and untrusted function
-     declarations *)
-  let ufuncs = List.map (fun f -> f.uf_fdecl) ec.ufunc_decls in
-  let tfuncs = List.map (fun f -> f.tf_fdecl) ec.tfunc_decls in
-  let funcs = List.append ufuncs tfuncs in
-  List.iter
-    (fun f ->
-      warn_non_portable_types f ;
-      warn_signed_size_or_count_types f ;
-      warn_size_and_count_params f )
-    funcs
-
-(** Includes are emitted in [args.h]. Imported functions have already
-    been brought into function lists. *)
-let gen_t_h (ec : enclave_content) (ep : edger8r_params) =
-  let oe_gen_tfunc_prototypes (tfs : trusted_func list) =
-    if tfs <> [] then
-      List.map (fun f -> sprintf "%s;" (oe_gen_prototype f.tf_fdecl)) tfs
-    else ["/* There were no ecalls. */"]
+(** Generate the Enclave code. *)
+let gen_enclave_code (ec : enclave_content) (ep : edger8r_params) =
+  (* Generate [args.h] which contains [struct]s for ecalls and ocalls *)
+  let oe_gen_args_header (ec : enclave_content) (dir : string) =
+    let oe_gen_marshal_struct (fd : func_decl) (errno : bool) =
+      let gen_member_decl (ptype, decl) =
+        let aty = get_param_atype ptype in
+        let tystr = get_tystr aty in
+        let tystr =
+          if is_foreign_array ptype then
+            sprintf "/* foreign array of type %s */ void*" tystr
+          else tystr
+        in
+        let need_strlen p =
+          (is_str_ptr p || is_wstr_ptr p) && (is_in_ptr p || is_inout_ptr p)
+        in
+        let id = decl.identifier in
+        [ [sprintf "%s %s;" tystr id]
+        ; (if need_strlen ptype then [sprintf "size_t %s_len;" id] else []) ]
+        |> List.flatten
+      in
+      let struct_name = fd.fname ^ "_args_t" in
+      let retval_decl = {identifier= "_retval"; array_dims= []} in
+      let members =
+        [ ["oe_result_t _result;"]
+        ; ( if fd.rtype = Void then []
+          else gen_member_decl (PTVal fd.rtype, retval_decl) )
+        ; (if errno then ["int _ocall_errno;"] else [])
+        ; flatten_map gen_member_decl (List.map conv_array_to_ptr fd.plist) ]
+        |> List.flatten
+      in
+      [ sprintf "typedef struct _%s" struct_name
+      ; "{"
+      ; "    " ^ String.concat "\n    " members
+      ; sprintf "} %s;" struct_name
+      ; "" ]
+    in
+    let oe_gen_user_includes (includes : string list) =
```

Comment:
I think I ought to refactor this back to their calling sites, as they're now much simpler. I've also soured on the explicit comments emitted when they would otherwise return empty. Let me know what you think.